### ISSUE TITLE
SuperTextField layout is wrong with custom text height

### ISSUE THREAD

**Issue by knopp** [superlistapp/super_editor#1864]
Video

https://github.com/superlistapp/super_editor/assets/96958/f516cca6-da66-4923-9f18-65b2babe58c0

You can see that the hint is aligned correctly while the actual text is not.

Modified snippet from `_single_line_demo.dart`

```dart
SuperDesktopTextField(
  key: _textKey,
  textController: _textFieldController,
  focusNode: _focusNode,
  decorationBuilder: (context, child) {
    return Container(
      decoration: BoxDecoration(
        borderRadius: BorderRadius.circular(4),
        border: Border.all(
          color: _focusNode!.hasFocus ? Colors.blue : Colors.grey.shade300,
          width: 1,
        ),
      ),
      child: child,
    );
  },
  hintBuilder: (context) {
    return const Text(
      'enter some text',
      style: TextStyle(
        color: Colors.grey,
        fontSize: 14,
        height: 2.0,
      ),
    );
  },
  hintBehavior: HintBehavior.displayHintUntilTextEntered,
  textStyleBuilder: (a) => demoTextStyleBuilder(a).copyWith(height: 2.0),
  minLines: 1,
  maxLines: 1,
),
```

**Comment by miguelcmedeiros** [comment]
Cc @matthew-carroll

**Comment by angelosilvestre** [comment]
Some updates on this. I increased the font height to `6.0` so the differences are move visible.

This is how it looks right now:

![SCR-20240306-sfju](https://github.com/superlistapp/super_editor/assets/7597082/774ee1a7-82de-481d-bbad-e4fd8d0bff96)

We can notice that the text isn't centered in relation to the caret.

I found a Flutter issue about vertically aligning text with custom font heights: https://github.com/flutter/flutter/issues/72521. This issue was closed after a PR which adds a `leadingDistribution` property to `TextStyle` landed.  After applying `TextLeadingDistribution.even` to the text style this is the output:

![SCR-20240306-shjc](https://github.com/superlistapp/super_editor/assets/7597082/8d9e1563-82c4-488a-a151-854a70233886)

Now we can notice that the text is centered in relation to the caret, but not centered in the viewport:

![image](https://github.com/superlistapp/super_editor/assets/7597082/01ffee86-fb51-4669-b5f6-df92a9944477)

It seems there is something wrong with our viewport size computation, because the height changes after typing the first character:

https://github.com/superlistapp/super_editor/assets/7597082/fea8c8dd-f756-48a8-88f0-43fb5552029d

While debugging I configured a fixed viewport height and the layout seems correct:

https://github.com/superlistapp/super_editor/assets/7597082/637c4409-b3a4-4f5f-b02e-a9604c9d8f1f

To summarize, apps need to configure the `TextStyle` like the following:

```dart
TextStyle textStyle = const TextStyle(
  leadingDistribution: TextLeadingDistribution.even,
  ...
);
```

And we need to figure out why the viewport height calculation is wrong. I'm looking into that.

**Comment by angelosilvestre** [comment]
@matthew-carroll Currently, we call `textLayout.getLineHeightAtPosition` to determine the viewport height. This method returns `character box height * line height`, which seems to be bigger than it should be.

Should we use the caret height instead? Using the caret height the layout looks correct:

https://github.com/superlistapp/super_editor/assets/7597082/cdbe3a55-a738-4dc3-a5d0-e6d9d5f405e9

**Comment by knopp** [comment]
@angelosilvestre, here is a repro for the golden we have failing:

[repro.tar.gz](https://github.com/superlistapp/super_editor/files/14653992/repro.tar.gz)

You can regenerate goldens with `flutter test --dart-define=CI=true --update-goldens`

It should result in something like this:
![editor_toolbar_url](https://github.com/superlistapp/super_editor/assets/96958/e1271422-cf5f-49d1-b0ed-0180b3ef3fde)

The yellow part is not supposed to be there.

**Comment by angelosilvestre** [comment]
@knopp The golden test is picking the android textfield, and it seems the mobile textfields needed the same fix. Could you please try the PR again?

**Comment by knopp** [comment]
I can confirm this fixes the golden. Thanks!
